app/jobs/autoscaling.py

```diff
@@ -181,26 +215,166 @@ def clamp_to_valid_range(asg, desired_capacity):
+
+
+def get_draining_servers(asg, tag_list, draining_tag):
+    ''' returns a map of draining instance IDs to the time they started draining '''
+    protected_instance_ids = [inst["InstanceId"] for inst in asg["Instances"] if inst["ProtectedFromScaleIn"]]
+    instance_dict = { item['ResourceId']: int(item['Value']) for item in tag_list if item['Key'] == draining_tag and item['ResourceId'] in protected_instance_ids }
+    return instance_dict
+
+
+def count_running_alignment_jobs(asg, tag_list, job_tag_prefix, job_tag_expiration):
+    ''' returns a map of instance IDs to number of jobs running on the instance '''
+    instance_ids = instances_in(asg)
+    count_dict = { id: 0 for id in instance_ids }
+    expired_jobs = []
+    for item in tag_list:
+        if item['Key'].startswith(job_tag_prefix):
+            job_tag = item['Key']
+            instance_id = item['ResourceId']
+            unixtime = int(item['Value'])
+            if unixtime_now() - unixtime < job_tag_expiration:
+                count_dict[instance_id] += 1
+            else:
+                expired_jobs.append(job_tag)
+    if expired_jobs:
+        print "The following job tags have expired and will be deleted: " + ", ".join(expired_jobs)
+        delete_tags_from_instances(instance_ids, expired_jobs)
+    return count_dict
+
+
+def delete_tags_from_instances(instance_ids, tag_keys):
+    if not instance_ids or not tag_keys:
+        return
+    cmd = "aws ec2 delete-tags --resources {list_instances} --tags {list_tags}"
+    cmd = cmd.format(list_instances=' '.join(instance_ids), list_tags=' '.join(['Key='+tag for tag in tag_keys]))
+    aws_command(cmd)
+   
+
+def remove_termination_protection(instance_ids, asg):
+    if not instance_ids:
+        return
+    cmd = "aws autoscaling set-instance-protection --instance-ids {list_instances} --auto-scaling-group-name {asg_name} --no-protected-from-scale-in"
+    cmd = cmd.format(list_instances=' '.join(instance_ids), asg_name=asg['AutoScalingGroupName'])
+    aws_command(cmd)
+
+
+def check_draining_servers(asg, tag_list, draining_tag, job_tag_prefix, job_tag_expiration, min_draining_wait, can_scale):
+    drain_date_by_instance_id = get_draining_servers(asg, tag_list, draining_tag)
+    num_jobs_by_instance_id = count_running_alignment_jobs(asg, tag_list, job_tag_prefix, job_tag_expiration)
+    instance_ids_to_terminate = []
+    current_timestamp = unixtime_now()
+    for instance_id, drain_date in drain_date_by_instance_id.items():
+        draining_interval = current_timestamp - drain_date
+        num_jobs = num_jobs_by_instance_id[instance_id]
+        if num_jobs == 0 and draining_interval > min_draining_wait:
+            instance_ids_to_terminate.append(instance_id)
+    if instance_ids_to_terminate:
+        print "The following instances are ready to be terminated: " + ", ".join(instance_ids_to_terminate)
+        if can_scale:
+            remove_termination_protection(instance_ids_to_terminate, asg)
```

Comment:
Do development environments share an ASG nowadays?   I don't think this is a major problem, but perhaps can_scale should be used to bypass most of the work here, not just be deferred to the end -- so development environments don't get bogged down doing AWS operations that are mute.